polus-unet-testing-plugin/src/unet_test.py

```diff
@@ -0,0 +1,171 @@
+
+    Preprocess image and run unet segmentation using caffe binary.
+
+    Args:
+        input_img: input image to be segmented.
+        img_pixelsize_x: input image pixel size in x.
+        img_pixelsize_y: input image pixel size in y.
+        weightfile_path: weight file path.
+        iofile_path: input/output file to save the preprocessed image and final segmentation score.
+
+    Returns:
+        Segmentation Mask.
+    """
+
+    #fixed parameters
+    n_iterations=0
+    tiling_x=4
+    tiling_y=4
+    modelfile_path = "2d_cell_net_v0-cytoplasm.modeldef.h5"
+    ## prepare image rescaling
+    np.set_printoptions(threshold=sys.maxsize)
+
+    #get model resolution (element size) from modelfile
+    modelfile_h5 = h5py.File(modelfile_path,'r')
+    modelresolution_y = modelfile_h5['unet_param/element_size_um'][0]
+    modelresolution_x = modelfile_h5['unet_param/element_size_um'][1]
+    modelfile_h5.close()       
+    #get input image absolute size
+    abs_size_x = input_img.shape[1] * img_pixelsize_x
+    abs_size_y = input_img.shape[0] * img_pixelsize_y
+    #get rescaled image size in pixel
+    rescaled_size_px_x = int(np.round(abs_size_x / modelresolution_x))
+    rescaled_size_px_y = int(np.round(abs_size_y / modelresolution_y))
+    rescale_size = (rescaled_size_px_x,rescaled_size_px_y)
+    ### preprocess image and store in IO file
+
+    #normalize image, then rescale
+    normalized_img = normalize(input_img)
+    rescaled_img = np.float32(rescale(rescale_size,normalized_img,mode='float32'))
+    #prepending singleton dimensions to get the desired blob structure
+    h5ready_img = np.expand_dims(rescaled_img, axis=(0,1))
+    iofile_h5 = h5py.File(iofile_path,mode='x')
+    iofile_h5.create_dataset('data',data=h5ready_img)
+    iofile_h5.close()
+
+    ### run caffe_unet commands
+    cmd = "caffe_unet tiled_predict -infileH5 "+iofile_path+" -outfileH5 "+iofile_path+" -model "+modelfile_path+\
+          " -weights "+weightfile_path+" -iterations "+str(n_iterations)+" -n_tiles "+str(tiling_x)+"x"+str(tiling_y)
+    os.system(cmd)
+    # load results from io file and return
+    output_h5 = h5py.File(iofile_path)
+    score = output_h5['score'][:]
+    output_h5.close()
+    #get segmentation mask by taking channel argmax
+    segmentation_mask = np.squeeze(np.argmax(score, axis=1))
+    return segmentation_mask
+
+
+def run_segmentation(inpDir, filePattern, pixelsize, weights, weightsfilename, outDir):
+    """Save segmentation mask using Biowriter.
+
+    Read input image, call segmentation function and save output mask using Biowriter.
+
+    Args:
+        inpDir: input image directory.
+        filePattern: input file name pattern to filter data.
+        pixelsize: image pixel size.
+        weights: weights file folder.
+        weightsfilename: weights file name.
+        outDir: output directory to save masks.
+    """
+
+    img_pixelsize_x = int(pixelsize)                 
+    img_pixelsize_y = int(pixelsize)
```

Comment:
Is there a way we can make `pixelsize` an optional input or detect the pixel size from the metadata? It would look something like this:

```python
img_pixelsize_x = img_pixelsize_y = None if pixelsize is None else int(pixelsize)
```

Inside of the for loop, inside of the `BioReader` block:

```python
# Set pixel size based on metadata from each image if it exists
if img_pixelsize_x is None and br.ps_x[0] is not None:
    img_pixelsize_x = br.ps_x[0]
    img_pixelsize_y = br.ps_y[0]
else:
    # Set to the model resolution so no resizing occurs
    img_pixelsize_x = model_resolution_x
    img_pixelsize_y = model_resolution_y
```

The reason to do this is that biologists don't always know/remember what the pixel resolution of their images are. You may also have a mixed collection of images with different resolutions. This code will make the pixel size parameter for your plugin an optional parameter.

The one thing you will have to consider is converting values to micrometers. I can share some additional code if that would be helpful.